src/replay/consensus.zig

```diff
@@ -0,0 +1,1321 @@
+        heaviest_fork_failures.items.len != 0)
+    {
+        // Implemented the log
+    }
+
+    // Vote on the fork
+    if (maybe_voted_slot) |voted| {
+        const slot_tracker = deps.slot_tracker;
+        var found_slot = slot_tracker.slots.get(voted.slot) orelse
+            return error.MissingSlot;
+
+        const voted_hash = found_slot.state.hash.read().get().* orelse
+            return error.MissingSlotInTracker;
+
+        try handleVotableBank(
+            deps.allocator,
+            deps.blockstore_reader,
+            voted.slot,
+            voted_hash,
+            deps.slot_tracker,
+            deps.replay_tower,
+            deps.progress_map,
+            deps.fork_choice,
+        );
+    }
+
+    // Reset onto a fork
+    if (maybe_reset_slot) |reset_slot| {
+        // TODO implement
+        _ = &reset_slot;
+    }
+}
+
+const LastVoteRefreshTime = struct {
+    last_refresh_time: sig.time.Instant,
+    last_print_time: sig.time.Instant,
+};
+
+/// Determines whether to refresh and submit an updated version of the last vote based on several conditions.
+///
+/// A vote refresh is performed when all of the following conditions are met:
+/// 1. Validator Status:
+///    - Not operating as a hotspare or non-voting validator
+///    - Has attempted to vote at least once previously
+/// 2. Fork Status:
+///    - There exists a heaviest slot (`heaviest_slot_on_same_fork`) on our previously voted fork
+///    - We've successfully landed at least one vote (`latest_landed_vote_slot`) on this fork
+/// 3. Vote Progress:
+///    - Our latest vote attempt (`last_vote_slot`) is still tracked in the progress map
+///    - The latest landed vote is older than our last vote attempt (`latest_landed_vote_slot` < `last_vote_slot`)
+/// 4. Block Progress:
+///    - The heaviest bank is sufficiently ahead of our last vote (by at least `REFRESH_VOTE_BLOCKHEIGHT` blocks)
+/// 5. Timing:
+///    - At least `MAX_VOTE_REFRESH_INTERVAL_MILLIS` milliseconds have passed since last refresh attempt
+///
+/// If all conditions are satisfied:
+/// - Creates a new vote transaction for the same slot (`last_vote_slot`) with:
+///   * Current timestamp
+///   * New blockhash from `heaviest_bank_on_same_fork`
+///   * Fresh signature
+/// - Submits this refreshed vote to the cluster
+///
+/// Returns:
+/// - `true` if a refreshed vote was successfully created and submitted
+/// - `false` if any condition was not met or the refresh failed
+///
+/// Note: This is not simply resending the same vote, but creating a new distinct transaction that:
+/// - Votes for the same slot
+/// - Contains updated metadata (timestamp/blockhash)
+/// - Generates a new signature
+/// - Will be processed as a separate transaction by the network
+///
+/// Analogous to [maybe_refresh_last_vote](https://github.com/anza-xyz/agave/blob/ccdcdbe9b6ff7dbd583d2101fe57b7cc41a6f863/core/src/replay_stage.rs#L2606)
+fn maybeRefreshLastVote(
```

Comment:
It's great that you already implemented this and wrote some extensive test coverage for it, because it's ultimately necessary for a full validator to execute this logic. 

For a non-voting validator, it seems that this function has no meaningful effects. Does that sound right to you?